File: examples/kmidtest/kmidtest.cpp
/*
    KMid2 MIDI/Karaoke Player
    Copyright (C) 2009-2010 Pedro Lopez-Cabanillas <plcl@users.sf.net>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include "kmidtest.h"
#include "kmidtest.moc"

#include <KMediaPlayer/View>
#include <KPluginFactory>
#include <KPluginLoader>
#include <KFileDialog>
#include <KStandardAction>
#include <KActionCollection>
#include <KMessageBox>
#include <KApplication>
#include <KUrl>
#include <KAboutData>
#include <KCmdLineArgs>
#include <KCmdLineOptions>
#include <KLocale>

KMidTest::KMidTest()
    : KParts::MainWindow()
{
    KPluginFactory *factory = KPluginLoader("kmid_part").factory();
    if (factory) {
        m_part = factory->create<KMediaPlayer::Player>(this);
        if (m_part) {
            setCentralWidget(m_part->view());
            setupActions();
            setupGUI();
            createGUI(m_part);
            //m_part->view()->setButtons(KMediaPlayer::View::All);
        }
    } else {
        KMessageBox::error(this, i18n("Could not find KMidPart!"));
        kapp->quit();
        return;
    }
}

KMidTest::~KMidTest()
{ }

void KMidTest::setupActions()
{
    KStandardAction::open(this, SLOT(fileOpen()), actionCollection());
    KStandardAction::quit(kapp, SLOT(closeAllWindows()), actionCollection());
    createStandardStatusBarAction();
    setStandardToolBarMenuEnabled(true);
}

void KMidTest::fileOpen()
{
    KUrl url = KFileDialog::getOpenUrl(KUrl(),QLatin1String("audio/midi"),this);
    if (!url.isEmpty())
        m_part->openUrl(url);
}

void KMidTest::play(KUrl u)
{
    m_part->openUrl(u);
}

static const char description[] =
    I18N_NOOP("A KDE KPart Application");

static const char version[] = "0.1";

int main(int argc, char *argv[])
{
    KAboutData about("kmidtest", 0, ki18n("KMidTest"),
                     version, ki18n(description), KAboutData::License_GPL,
                     ki18n("(C) 2010 Pedro Lopez-Cabanillas"),
                     KLocalizedString(), 0, "plcl@users.sf.net" );
    about.addAuthor( ki18n("Pedro Lopez-Cabanillas"), KLocalizedString(),
                     "plcl@users.sf.net" );
    KCmdLineArgs::init(argc, argv, &about);
    KCmdLineOptions options;
    options.add("+[url]", ki18n( "Song to open" ));
    KCmdLineArgs::addCmdLineOptions(options);
    KApplication app;

    if (app.isSessionRestored())
        RESTORE(KMidTest)
    else {
        KMidTest *widget = new KMidTest;
        KCmdLineArgs *args = KCmdLineArgs::parsedArgs();
        if (args != 0 && args->count() > 0)
            widget->play(args->url(0));
        widget->show();
    }
    return app.exec();
}

File: examples/kmidtest/kmidtest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
#   KMid2 MIDI/Karaoke Player
#   Copyright (C) 2009-2010 Pedro Lopez-Cabanillas <plcl@users.sf.net>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License along
#   with this program; if not, write to the Free Software Foundation, Inc.,
#   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#

from PyKDE4.kdecore import *
from PyKDE4.kdeui import *
from PyKDE4.kio import *
from PyKDE4.kparts import *
from PyQt4.QtCore import *
import sys


class MainWindow (KParts.MainWindow):
    def __init__ (self, app):
        KParts.MainWindow.__init__(self)
        self.app = app
        self.resize(330, 130)
        factory = KPluginLoader("kmid_part").factory()
        if (factory):
            part = factory.create(self, "KMidPart")
            if (part):
                self.part = part
                self.setCentralWidget(part.widget())
                self.act('file_open', KStandardAction.Open, self.openDocument)
                self.act('file_quit', KStandardAction.Quit, self.app.quit)
                self.setStandardToolBarMenuEnabled(True)
                self.setupGUI()
                self.createGUI(part)
        
    def openDocument(self):
        url = KFileDialog.getOpenUrl(KUrl(), "audio/midi", self, "Open File")
        if not url.isEmpty():
            self.part.openUrl(url)
            
    def act(self, name, texttype, func):
        a = self.actionCollection().addAction(texttype, name)
        QObject.connect(a, SIGNAL("triggered()"), func)
        return a

#--------------- main ------------------
if __name__ == '__main__':
    appName     = "kmidtest"
    catalog     = ""
    programName = ki18n("KMid Part Example")
    version     = "0.1"
    description = ki18n("Example loading KMidPart")
    license     = KAboutData.License_GPL
    copyright   = ki18n("(c) 2010 Pedro Lopez-Cabanillas")
    text        = ki18n("KDE4 MIDI/Karaoke player")
    homePage    = "http://kmid2.sourceforge.net"
    bugEmail    = "plcl@users.sf.net"
    
    aboutData   = KAboutData(appName, catalog, programName, version, description,
                                license, copyright, text, homePage, bugEmail)
           
    KCmdLineArgs.init(sys.argv, aboutData)
        
    app = KApplication()
    mainWindow = MainWindow(app)
    mainWindow.show()
    app.exec_()
